src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.TryGet.cs

```diff
@@ -490,8 +535,53 @@ public bool TryGetDateTimeOffset(out DateTimeOffset value)
                 throw ThrowHelper.GetInvalidOperationException_ExpectedString(TokenType);
             }
 
-            ReadOnlySpan<byte> span = HasValueSequence ? ValueSequence.ToArray() : ValueSpan;
-            return JsonHelpers.TryParseAsISO(span, out value, out int bytesConsumed) && span.Length == bytesConsumed;
+            ReadOnlySpan<byte> span = ValueSpan;
+
+            if (HasValueSequence)
+            {
+                Span<byte> stackSpan;
+
+                int sequenceLength = (int)ValueSequence.Length;
+                if (sequenceLength <= JsonConstants.StackallocThreshold)
+                {
+                    // Cannot create a span directly since it gets passed to instance methods on a ref struct.
+                    unsafe
+                    {
+                        byte* ptr = stackalloc byte[sequenceLength];
+                        stackSpan = new Span<byte>(ptr, sequenceLength);
+                    }
+                }
+                else
+                {
+                    stackSpan = new byte[sequenceLength];
+                }
+
+                ValueSequence.CopyTo(stackSpan);
+                span = stackSpan;
+            }
+
+            int bytesConsumed;
+
+            if (_stringHasEscaping)
+            {
+                int idx = span.IndexOf(JsonConstants.BackSlash);
+                Debug.Assert(idx != -1);
+
+                Span<byte> utf8Unescaped = (span.Length <= JsonConstants.StackallocThreshold)
+                    ? stackalloc byte[span.Length]
+                    : new byte[span.Length];
```

Comment:
Either we should use array pool for the intermediary buffer for unescaping (that is beyond `StackallocThreshold`), or we should constrain the input (similar to `TryGetGuid`) such that stackalloc is fine or for all cases.

Reply:
We now constrain the input such that stackalloc is fine for all cases.